node/subsystem-util/src/lib.rs

```diff
@@ -209,6 +210,53 @@ specialize_requests! {
 	fn request_validation_code_hash(para_id: ParaId, assumption: OccupiedCoreAssumption)
 		-> Option<ValidationCodeHash>; ValidationCodeHash;
 	fn request_on_chain_votes() -> Option<ScrapedOnChainVotes>; FetchOnChainVotes;
+	fn request_session_executor_params(session_index: SessionIndex) -> Option<ExecutorParams>; SessionExecutorParams;
+}
+
+/// Requests executor parameters from the runtime effective at given relay-parent. First obtains
+/// session index at the relay-parent, relying on the fact that it should be cached by the runtime
+/// API caching layer even if the block itself has already been pruned. Then requests executor
+/// parameters by session index.
+/// Returns an error if failed to communicate to the runtime, or the parameters are not in the
+/// storage, which should never happen.
+/// Returns default execution parameters if the runtime doesn't yet support `SessionExecutorParams`
+/// API call.
+/// Otherwise, returns execution parameters returned by the runtime.
+pub async fn executor_params_at_relay_parent(
+	relay_parent: Hash,
+	sender: &mut impl overseer::SubsystemSender<RuntimeApiMessage>,
+) -> Result<ExecutorParams, Error> {
+	match request_session_index_for_child(relay_parent, sender).await.await {
+		Err(err) => {
+			// Failed to communicate with the runtime
+			Err(Error::Oneshot(err))
+		},
+		Ok(Err(err)) => {
+			// Runtime has failed to obtain a session index at the relay-parent; should never happen
+			Err(Error::RuntimeApi(err))
+		},
+		Ok(Ok(session_index)) => {
+			match request_session_executor_params(relay_parent, session_index, sender).await.await {
+				Err(err) => {
+					// Failed to communicate with the runtime
+					Err(Error::Oneshot(err))
+				},
+				Ok(Err(_)) => {
+					// Runtime doesn't yet support the api requested, should execute anyway
+					// with default set of parameters
+					Ok(ExecutorParams::default())
+				},
+				Ok(Ok(None)) => {
+					// Storage doesn't contain a parameter set for the given session; should
+					// never happen
+					Err(Error::RuntimeApi(RuntimeApiError::NotSupported {
```

Comment:
This is a bit misleading. `NotSupported` is usually used when the runtime doesn't contain the function (my comment above). Here the runtime call is executed successfully but the runtime doesn't contain a value. 

I think it makes more sense to return `Ok(None)` and handle this in the caller. That way he/she will know that the call was successful but returned no data.

Reply:
I thought about that, and I'm not sure it makes much sense. The caller is requesting execution environment parameters. The absence of the parameters is not a normal situation; it's an error from the caller's perspective. We are supposed to return either stored parameters or default parameters if the runtime doesn't support stored parameters yet. `Ok(None)` is useless for the caller anyway; it doesn't have any choice besides processing it as the same error when the state is pruned, for example. So returning `Ok(None)` here will just force the caller to extend the `Err(_) =>` match to `Err(_) | Ok(None) =>` match.

But you're right that `NotSupported` is dubious; I'll change it to `RuntimeApiError::Execution` (I just need to figure out what error it should wrap).

This comment also reminded me about a transitional period problem: if a runtime upgrade is performed in the middle of a session, we don't have execution environment parameters for that session in the storage, but still, the runtime does support execution parameters after the upgrade, and all the attempts to execute  PVFs in the same session would fail. I'll address it in a migration.

Reply:
Okay, I see your point. It's pointless to return `Ok(None)`.
`RuntimeApiError::Execution` will be better choice in this case. You can even wrap a custom error inside it.

Reply:
Here's the problem: `RuntimeApiError::Execution` is supposed to be sourced from another error. Yes, I can still create that inner error myself and provide it as a source, but I believe it would be a poor design: `RuntimeApiError::Execution` is meant to generalize underlying errors, not produce its own ones. So, for now, I've added the `subsystem_util::Error::DataNotAvailable` variant. Let me know if you see any problems with that.

Reply:
I think it's fine. In your case the data is mandatory but the API always return an option. It's worth having a type representing 'Mandatory data in version X is missing'.

Reply:
On the other side - for the runtime this is pretty normal - the value can be there or cannot be there. The calling subsystem is the one knowing that the value is mandatory and must present in the runtime. I still believe in my initial thought to handle missing data at the caller, but I don't insist on it. Having special error for such cases is also sensible.

Reply:
These discussions around the design are immensely useful; for example, this one brought up one significant concern I forgot about.

Why, in the first place, has it been designed to return `Option` and not just `ExecutorParams` if they should always be there? Because they shouldn't. If the runtime is upgraded in session 2, we have parameters from session 2 but don't have them for session 1. Then, if a candidate from session 1 is disputed, the runtime will legitimately return `None`. And in that case, it's incorrect to return an error to the caller, we should return a default parameter set for the PVF to be able to execute as in the pre-parametrized environment.

@eskimor @ordian I see several design options here:
1) To store the index of the first session for which execution parameters are available and use the default ones if parameters are requested for a session with a lower index. It looks like the "right" way, but if stored on-chain, it would require one more storage unit, one more runtime call, one more cache, and so on, so it complicates things;
2) like 1) but everything is sorted out on the runtime side, that is, the first session index is stored on-chain, and inside the `session_executor_params()` runtime call, runtime implementation code checks for the session index and returns default parameters if it's lower than the first session index. That makes life easier, but I remember @rphmeier's concern that the runtime implementation code should be kept as simple as calling underlying functions and not involve any complex logic; 
3) During the migration, insert the default parameter set into the storage not only for the current session but also for the `dispute_period` sessions back. Kinda hacky, but we only need it for the first `dispute_period` anyway, and any code assigned to sort this situation out would become useless after 6 sessions, so it seems legit to put it into the migration.

Which one do you guys think is better? (If nobody answers, I'll take number 3).

Reply:
4. Can we upgrade the runtime at least 6 sessions before the client side changes will be released? That is a separate storage migration before adding a runtime api call.

I think option 3 sounds reasonable.

Reply:
Hmm, it is either supported or not. If it is not yet supported in a session - we get the not (yet)supported error and use default parameters.

So why do we need `None` - we would ask the version 1 runtime for the session not supporting it, so we would just get unsupported and use default. I think, I am missing something.

We are calling the runtime in the context of a block and we will execute the runtime version as it was at this block.

Reply:
oh wait, we would not necessarily call the runtime at that block I guess: So we would look into the state of a current block, sounds like a job for a migration to me. So yes option 3: Not hacky, it is just a necessary migration. We should then also be able to get rid of the `None`.

Reply:
FYI option 3 was implemented already.

Reply:
Today's testing on Versi has revealed an interesting problem connected to this discussion.

Say the runtime upgrade introducing executor environment parametrization happens at block 95, session 30. Migration runs and puts default execution parameters to the storage for session indexes 24..30. But those changes are made for the state at block 95.

Then, we're validating a candidate from block 94. We request the session index at block 94 (it's 30), and then we're requesting execution parameters for session index 30 **at block 94**, and they're obviously missing. Thus, the validator starts to claim disputes against a valid candidate.

The aforementioned variants 1, 2, and probably 4 may solve that. But I wouldn't say I like 1 and 2, and I'm not sure how much effort is needed to proceed with 4.

So I propose a variant number 5: if execution parameters are unavailable for any reason, always return default ones. That will bootstrap parametrization itself. After some time, we'll release a new node software version that takes care of returned errors. Our only concern is not to change any executor environment parameters until the latter node software is deployed to the vast majority of validators.

@eskimor @ordian does it make any sense?

Reply:
@eskimor you also mentioned that we could always call the runtime at the current block instead of the candidate's parent. That would also solve the issue, but I don't know if it's a proper solution.

Reply:
I'm confused a bit. Why wouldn't API version at block 94 be below the one that introduced the runtime api call returning NotSupported error?
```rust
				// Runtime doesn't yet support the api requested, should execute anyway
				// with default set of parameters
```

Reply:
@ordian a good point! You're right, in a real environment, it should work like that and so it wouldn't be a problem. The only problem is testing. I need to figure out how to proceed with testing and overcome those obstacles.

Reply:
bump the ee runtime api to version 4?

Reply:
or have it at version 2 before?